### ISSUE TITLE
IOOS Metadata checker needs Wicken updating

### ISSUE THREAD

**Issue by daf** [ioos/compliance-checker#216]
There aren't any tests for it, and netcdf files work fine for it, so it wasn't caught during the Wicken removal.  The XML paths need updating.

**Comment by lukecampbell** [comment]
Didn't we can wicken?

**Comment by daf** [comment]
We did, but the IOOS Metadata checker isn't fully updated for that change, specifically in non-netcdf cases.

**Comment by lukecampbell** [comment]
Even though I stuck it in the milestone, I'm probably going to use the resources to REMOVE wicken completely.

**Comment by lukecampbell** [comment]
XPath should suffice, I don't need this weird XML ORM thing. Plus with all the work that @cmorse, @abirger and @kwilcox have done to manage the mappings between SOS and NetCDF, we can update the IOOS Metadata Checker.

**Comment by daf** [comment]
^^ all of this

**Comment by cmorse** [comment]
@lukecampbell Pretty sure I'm not the person you meant to @ mention.

**Comment by lukecampbell** [comment]
Yeah sorry I meant @cheryldmorse

**Comment by benjwadams** [comment]
I'm not seeing wicken when I run `pip freeze` on my virtualenv.

**Comment by benjwadams** [comment]
I don't see Wicken when I run `pip freeze` in my virtualenv.  I'm also not entirely clear on the issue here -- is it that some checks on the IOOS suite (`-t ioos`) are not working properly on non-NetCDF datasets as there are no XPaths?

**Comment by daf** [comment]
Right - the IOOS checker won't run on GetCapabilities/DescribeSensor documents as that codepath was never updated to work without Wicken.

**Comment by benjwadams** [comment]
Do you know of a last known good commit WRT the IOOS SOS check?  Tried bisecting and there are just a lot of missing fields, not outright failures.

**Comment by daf** [comment]
The last "working" commit would be pre-wicken removal, c4387985a15e578141bf15298067ab67e036af7d

I can't test how it works currently because I can't get past a bunch of random issues, see fixes in #309, and now I'm on some encoding issues with the XML document stuff.  I'm testing via:

```
~/dev/compliance-checker -> python cchecker.py --test ioos http://sos.maracoos.org/stable/sos/sldmb/43493/43493_2014-03-30.nc                             [fix-xml-urls]&cc-modern
Running Compliance Checker on the dataset from: http://sos.maracoos.org/stable/sos/sldmb/43493/43493_2014-03-30.nc
Traceback (most recent call last):
  File "cchecker.py", line 79, in <module>
    sys.exit(main())
  File "cchecker.py", line 68, in main
    args.format)
  File "/home/daf/dev/compliance-checker/compliance_checker/runner.py", line 44, in run_checker
    ds = cs.load_dataset(ds_loc)
  File "/home/daf/dev/compliance-checker/compliance_checker/suite.py", line 583, in load_dataset
    ds = self.process_doc(doc)
  File "/home/daf/dev/compliance-checker/compliance_checker/suite.py", line 485, in process_doc
    xml_doc = ET.fromstring(str(doc))
  File "src/lxml/lxml.etree.pyx", line 3213, in lxml.etree.fromstring (src/lxml/lxml.etree.c:76702)
  File "src/lxml/parser.pxi", line 1825, in lxml.etree._parseMemoryDocument (src/lxml/lxml.etree.c:115048)
ValueError: Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.
```

so I'm stopped up there (on python3 obviously).

**Comment by daf** [comment]
https://github.com/ioos/compliance-checker/blob/master/compliance_checker/base.py#L75-L86

MultipleXMLDogma refs are not imported anymore (they're from wicken), these codepaths must not be taken or they'd crash you.  Should be removed as part of this issue.

Does `@check_has` decorator work with XML documents?  https://github.com/ioos/compliance-checker/blob/master/compliance_checker/base.py#L237-L253

**Comment by benjwadams** [comment]
In theory the check_has should work, as it will usually end up called `std_check` or `std_check_in` which are Python attribute checks.  Provided the properly attributes are set on the checked object, it should work, but right now it's just using the OWSLib SOS object.
https://github.com/ioos/compliance-checker/blob/master/compliance_checker/base.py#L150-L170

**Comment by benjwadams** [comment]
OK @daf, the push I have above to my branch should hopefully take care of things.  I still need to make some unit tests against a mock SOS service so we avoid any future regressions for the IOOS test, so I'm going to do that hopefully tomorrow before I send in a formal PR.  Also haven't gotten to check Py3 but will do tomorrow.

I more or less blackbox tested against the prior behavior by running:

`git worktree add ioos_old be0a7fb && cd ioos_old && mkvirtualenv cchecker_old && pip install numpy==0.6 && pip install -r requirements.txt`

Compared the outputs from the pushed branch and the known good and the output appears the same, at least for the following URLs:

IOOS GetCaps:
http://data.oceansmap.com/thredds/sos/caricoos_ag/VIA/VIA.ncml

IOOS DescribeSensor:
http://data.oceansmap.com/thredds/sos/caricoos_ag/VIA/VIA.ncml?request=describesensor&service=sos&procedure=urn:ioos:station:ncsos:VIA&outputFormat=text/xml%3Bsubtype%3D%22sensorML/1.0.1/profiles/ioos_sos/1.0%22&version=1.0.0

Also, I'm not sure where the `altitude_units` pops up in the GetCapabilities. it appears to be in the old NCML mapping JSON file used by wicken -- perhaps it was added in error?

**Comment by benjwadams** [comment]
Best emoji in URL.
